### ISSUE TITLE
Negation with overflow possible in VarInt64 Deserialize

### ISSUE THREAD

**Issue by pchickey** [NikVolf/parity-wasm#92]
I just hit a panic for 'attempt to negate with overflow' when deserializing a binary, the root cause was in the negation at the following location:

https://github.com/NikVolf/parity-wasm/blob/a4e2ec0adb7bf39da9558681f87c32095827c768/src/elements/primitives.rs#L303

This occurs when `shift = 63`.

I think you want to treat the value as unsigned until the bits are all populated, and then cast it to signed. Could the `VarUint64` deserialize be reused here?

This issue might affect more of the VarInt deserializers as well.

**Comment by NikVolf** [comment]
Oh thanks for reporting!
Looks like you are right, can you attach wasm file (or part of it) to look closer though?

**Comment by NikVolf** [comment]
https://github.com/NikVolf/parity-wasm/pull/93

taking in account that this tests are ok, i'm still unsure where the problem is, wasm binary would be great

**Comment by pchickey** [comment]
I bisected down to the following wasm module. I used the `wast2wasm` tool from the `WebAssembly/wabt` repo to translate it to binary.

```
(module
(type (;0;) (func))
(func (;0;) (type 0)
i64.const -1152894205662152753
return))
```

I confirmed this by adding it to your test suite in my fork: https://github.com/pchickey/parity-wasm/commit/3ea668cd1d602be13a671abc743d8435b3d8681a

**Comment by pchickey** [comment]
I was curious if I could use a fuzzer to find the same bug, and once setup, I was able to in under a minute of compute time: https://gist.github.com/pchickey/20702a3842a71c4ff95fe6d8dbf0cf9b

The `wasm-opt` tool has a `--translate-to-fuzz` mode that takes an arbitrary input file (as a seed) and generates a valid wasm file. I used that inside a simple `cargo fuzz` harness: https://github.com/pchickey/parity-wasm/commit/3acf4516c89821ca5d5e553a407230adc59e0a35

The fuzzer requires a nightly rustc and the `cargo-fuzz` package installed via cargo. Happy to make a PR so you can run it yourself!

**Comment by NikVolf** [comment]
Ok, i think fix will come shortly, thanks

That's great suggestion about fuzz, i will make a separate issue for that
The PR will be great

#94

**Comment by NikVolf** [comment]
Fixed in #94
